samples/confidential_ml/src/service_app.py

```diff
@@ -42,17 +42,38 @@ def evaluate_sample():
         output = alexnet.evaluate_formatted(name)
         return f"file inferenced successfully.\n{output}"
 
-def download_and_decrypt(url, path, key):
-    output = io.BytesIO()
-    cl = pycurl.Curl()
-    cl.setopt(cl.URL, url)
-    #cl.setopt(pycurl.VERBOSE, 1)
-    cl.setopt(pycurl.FOLLOWLOCATION, 1)
-    cl.setopt(pycurl.WRITEFUNCTION, output.write)
-    cl.perform()
-    cl.close()
+def download_and_decrypt(url, path, key, max_retries=3):
+    # enumerate 
+    for i in enumerate(range(max_retries)):
+        try:
+            output = io.BytesIO()
+            cl = pycurl.Curl()
+            cl.setopt(cl.URL, url)
+            #cl.setopt(pycurl.VERBOSE, 1)
+            cl.setopt(pycurl.FOLLOWLOCATION, 1)
+            cl.setopt(pycurl.WRITEFUNCTION, output.write)
+            cl.perform()
+            cl.close()
+            if output.getvalue():
+                # aes_obj.decrypt() expects input to be a multiple of 16 in length
+                # because AES encrypts in blocksize of 16 bytes.
+                if len(output.getvalue()) % 16 != 0:
```

Comment:
Just curious, is this to heuristic to prevent download errors?

Reply:
Yes, there are two cases that are handled here:
1. (transient) network issues, which results in pycurl.error and will be retired up to `max_retries`
2. A successful request but an unexpected response is returned. For example, Azure storage API can return an error instead of the file.